api/scpca_portal/test/models/test_job.py

```diff
@@ -66,6 +100,157 @@ def test_submit_job_failure(self, mock_batch_submit_job):
+        self.assertTrue(success)  # Synced and updated the db
+
+        # Job should be updated and saved with correct field values
+        saved_job = Job.objects.get(pk=submitted_job.pk)
+        self.assertEqual(saved_job.state, JobStates.TERMINATED.name)
+        self.assertIsInstance(saved_job.terminated_at, datetime)
+        self.assertIsNone(saved_job.failure_reason)
+
+        # Set up mock for get_jobs with FAILED
+        submitted_job.state = JobStates.SUBMITTED.name
+        mock_batch_get_jobs.return_value = [
+            {
+                "status": "FAILED",
+                "statusReason": "Job FAILED",
+            }
+        ]
+
+        success = submitted_job.sync_state()
+        mock_batch_get_jobs.assert_called()
+        self.assertTrue(success)  # Synced and updated the db
+
+        # Job should be updated and saved with correct field values
+        saved_job = Job.objects.get(pk=submitted_job.pk)
+        self.assertEqual(saved_job.state, JobStates.COMPLETED.name)
+        self.assertEqual(saved_job.failure_reason, "Job FAILED")
+        self.assertIsInstance(saved_job.completed_at, datetime)
+
+    @patch("scpca_portal.batch.get_jobs")
+    def test_bulk_sync_state(self, mock_batch_get_jobs):
+        # Set up mock for get_jobs with all AWS Batch job statuses including terminated FAILED job
+        mock_batch_get_jobs.return_value = self.bulk_create_mock_aws_batch_jobs(
+            [
+                {"jobId": f"{self.mock_batch_job_id}-{i}", "status": status}
+                for i, status in enumerate(
+                    self.aws_batch_job_status,
+                )
+            ],
+            terminated_job={"jobId": f"{self.mock_batch_job_id}-7", "status": "FAILED"},
+        )
+        self.bulk_create_mock_jobs(
+            [
+                {"batch_job_id": f"{self.mock_batch_job_id}-{i}", "state": JobStates.SUBMITTED.name}
+                for i in range(0, 8)
+            ]
+        )
+
+        success = Job.bulk_sync_state()
+        mock_batch_get_jobs.assert_called_once()
+        self.assertTrue(success)
+
+        # Job with state change should be updated and saved with correct field values
+        for saved_job in Job.objects.all():
+            if saved_job.state == JobStates.TERMINATED.name:
+                self.assertIsNone(saved_job.failure_reason)
+                self.assertIsInstance(saved_job.terminated_at, datetime)
+            elif saved_job.state == JobStates.COMPLETED.name:
+                self.assertEqual(saved_job.state, JobStates.COMPLETED.name)
+                self.assertIn(saved_job.failure_reason, [None, "Job FAILED"])
+                self.assertIsInstance(saved_job.completed_at, datetime)
+            else:
+                self.assertEqual(saved_job.state, JobStates.SUBMITTED.name)
+                self.assertIsNone(saved_job.failure_reason)
+                self.assertIsNone(saved_job.completed_at)
+                self.assertIsNone(saved_job.terminated_at)
```

Comment:
This would probably be better if we pulled it outside of the same loop.

Instead of doing one query for all of the jobs at the same time you can query them individually and then do a different loop for each expected job.state. You can remove the assertEqual since it will be part of the query.

The last query instead of doing `Job.objects.filter(state=JobStates.SUBMITTED.name)` you could do `Job.objects.exclude(terminated_jobs + completed_jobs)`. This assumes you cached the new queryset for terminated jobs in `terminated_jobs` variable.